sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -133,37 +133,50 @@
   private final ValueProvider<Integer> numShardsProvider;
   private final boolean windowedWrites;
   private int maxNumWritersPerBundle;
+  // If set,this is used instead of DynamicDestinations.getSideInputs. This is used by the Beam
+  // translation code.
+  private final List<PCollectionView<?>> overriddenSideInputs;
+
+  private List<PCollectionView<?>> getSideInputs() {
+    return (overriddenSideInputs != null)
+        ? overriddenSideInputs
+        : getSink().getDynamicDestinations().getSideInputs();
```

Comment:
Certainly, you should remove the comment and make the field not nullable and just set it to the correct value in the constructor.

Beyond that, I'm not simply asking you to make a small tweak to the code, but explaining background that should influence but not dictate what you do. It is deeply suspicious that this works differently from every other place we use side inputs, so you should understand it well enough to be able to explain why it is OK. It may be, and maybe other places we could choose this design as well... but maybe not.